### ISSUE TITLE
New TTT Role Priest

### ISSUE THREAD

**Issue by h4ppy1337** [iViscosity/ttt2-role_priest#1]
Hello hello, as I said I tested the new Priest role on a private server to see if it works. First things first, in the priest/shared.lua there's a ")" missing in line 110, since there are 8 open brackets and only 7 closing ones.
When I add the missing bracket, the priest gets implemented as a role with his holy deagle.
Now, let's get started with all the features:

When shooting the Detective with the deagle, he loses some of his HP. This seems to work.

When shooting a Jackal, SK or normal traitor with the deagle, I die. This seems to work.
Anyway, when shooting a Hitman or Vampire with the deagle, nothing happens.

When shooting the Jester, he dies, but also wins the round. Do I have to change any of his winstates to prevent that?

When shooting an Innocent with the deagle, nothing happens. My server console gives back some information error code (I could show you in a screenshot) and the innocent does not get converted into brotherhood.
But, when I force a second priest to join the round, they can see each other in the player list. This information system seems to work.

When shooting the Sidekick with the deagle, he dies. This seems to work.

When shooting an Infected wit the deagle, he dies. This seems to work.

When infecting a Priest, no other priest gets infected. Could not test this with possible secondary brotherhood members, since converting does not work. But they probably also won't get infected (?)

The addon pretty much is already in shape, the priest spawns properly and the deagle is implemented. These are the things I tested in private, hope it helps you to improve the addon.

I also do have two questions to ask. First of all, I already found a way to change the priest's percentage in your code, but i did not find any way to randomize it. Is there any possibility to give the Priest a sort of 20% appearance ratio or something like that? Since that does already work with all the other roles, I guess it has to be possible to implement here?

Second of all, is there a way to give the priest some visual or audible clue, for when the holy deagle is charged? Since the ammunition bar does not get empty when fired, you have no idea when it's refilled.

I know, this is a lot of information at once, so just take your time, as for me you already did pretty much for all of this :)

**Comment by iViscosity** [comment]
I would like to see that error. It's likely (very likely) that it's the reason conversion doesn't work.

Jester winstates I forgot about, I will add that feature in.

Hitman and Vampire I forgot about.

Infecting the Brotherhood **should've** worked, but I'll look into that as well.

Deagle ammo bar should show as empty, and I also forgot to add a feature to give it more ammo after 60 seconds.

**Comment by h4ppy1337** [comment]
https://imgur.com/z3yroOt that's the error which occurs when trying to convert an innocent, strangely it did not occur when trying to convert Clairvorant and Survivalist. They still just did not get into brotherhood anyway.

Infecting the Brotherhood _might_ work. As I said, I had no secondary members to test this with, since conversion didn't work. For what I did, I forced two priests into the game, infecting one of them while the other one remained a Priest. But as there should only be a max amount of 1 priest in the game, this situation will most likely never show up in any real game. I don't know how far this "bug" goes in real game.

Deagle ammo bar shows as 1+00, even after shooting, never going empty, but having a cooldown of approx. 60 seconds after every shot. In addition to that it also does not shoot bullets, has no shot animation and no sound (not like the sidekick deagle, which acts like a normal deagle in physics and sound, except for the non-existing dmg and the side effect). It only shoots when I aim at someone and I only recognize I shot someone, because one of us dies, or if not, because you can also see a short blood effect where I shot him. Anyway, all of this seems to go a lot further from what you are sure about having coded, so maybe I fucked something up by copying the weapon files. But since that's not that hard of a thing to do, I have to guess I didn't. Just can't do anything but report these things to you, I guess if there are missing parts in the code, you're gonna find them.

**Comment by h4ppy1337** [comment]
oh, and something else i noticed in your recent updated code: Hitman and Vampire should not be killed, but be treated as a normal traitor instead, meaning that shooting them with the holy deagle should kill the Priest instead.

**Comment by iViscosity** [comment]
Looking at the [code for Jester](https://github.com/TTT-2/ttt2-role_jes/blob/master/gamemodes/terrortown/entities/entities/jester/shared.lua#L127), it seems that Jester will immediately win only if Jester winstate is 0 (meaning all other winstates are 0). I don't think that should be changed (it can be done, though).

**Comment by h4ppy1337** [comment]
So we should let the Jester win when getting shot with the Holy Deagle? It won't harm the priest's role much, so I wouldn't bother, that's for sure. Did you figure out what the problem with conversion was?

**Comment by iViscosity** [comment]
Yes I did. Forgot to do it, though. Turns out it was a problem with the scoreboard colour. When I made the custom team it makes a global variable (TEAM_BROTHERHOOD) that I didn't realize only pointed to the NAME and not the entire role data.

**Comment by h4ppy1337** [comment]
Ah, I see. I'm gonna test it out later and report it to you

**Comment by iViscosity** [comment]
I will have to do some experimentation on the problem with the short distance. I stole the code for getting a target from the SK's knife, and when I did it I realized that that might cause issues with the range. I'm gonna look into seeing how I can increase the range since I don't know really how :P

**Comment by h4ppy1337** [comment]
yeah, no problem dude. take your time with these things, since they probably still need much effort to fix properly. I'll have an eye on this, so hit me up anytime when I can help you or to test new features :)

**Comment by iViscosity** [comment]
New update posted. Should fix range issue.

**Comment by h4ppy1337** [comment]
Tried to test it, sadly shooting does not work :( Error code here: https://imgur.com/a/VV62Xmk

**Comment by h4ppy1337** [comment]
The above update fixed shooting with the Holy Deagle. The range now is extended and should be working properly now.

Sadly, the Priest is unable to drop any weapon now (this error code occurs: https://imgur.com/a/Hmn4cUb)
In addition to that, the effects of the Holy Deagle have been overwritten, since now everyone just gets 30HP of damage, no matter which role he has.

**Comment by iViscosity** [comment]
Those fixes fix the errors on dropping weapons. Will check about the effects now.

**Comment by h4ppy1337** [comment]
Dropping errors are fixed now, here's what remains from the above bugs:

- Survivalist and Clairvoyant can be seen by the Priest

- Holy Deagle does not trigger anything when shooting someone. It just fires, the bullet has no effect.

- Some short error code that showed in my console (idk what it belongs to since there was no visible bug related to the message occuring): https://imgur.com/a/n7APM06

**Comment by iViscosity** [comment]
Still nothing?

Weird...

**Comment by h4ppy1337** [comment]
With the current update the bullets have some influence on the game, but for now it's just that everyone takes 30HP damage from the deagle, no matter which role he has. 
I noticed that dead bodies don't get dropped. Maybe a side effect of the dropping mechanic? I don't know. But I got that error code from time to time in my console: https://imgur.com/a/i9AVtYh

**Comment by iViscosity** [comment]
Okay, latest patch added some debug statements. Weapon will be able to drop until I find a better way to fix that problem.

When shooting Infected, Sidekick, or Jester, a message saying "INF/SK/JES" should show up in the SERVER console.

I did similar things for the other roles.

Clairvoyant and Survivalist weren't edited at all, are those roles normally able to be seen by innocents?

**Comment by h4ppy1337** [comment]
Okay, I'm gonna test it out later. Clairvoyant and Survivalist are not supposed to be seen by other players. Strangely enough, they can't be seen by any player except for Priests now either. It has to have something to do with the Brotherhood team then, if I had to guess.

**Comment by iViscosity** [comment]
Any update with the debug messages?

**Comment by iViscosity** [comment]
Damn autoclosing :P

That will fix the latest error.

**Comment by h4ppy1337** [comment]
aight, lemme check this out

**Comment by h4ppy1337** [comment]
So the bug is fixed, but I can't see any statement in the console when shooting players. 
I also found these two bad boys here when scrolling through my console, although I don't know what they belong to: https://imgur.com/a/YgjK6Nc

**Comment by h4ppy1337** [comment]
Any update on the statements yet?

**Comment by iViscosity** [comment]
Sorry about being inactive. Strange how you don't see any of the statements... both of them originate from this addon but only the second one has to do with the EntityTakeDamage hook when I do my checks...

I've fixed both of those errors in next commit but strange how those messages don't show up... It must be the hook isn't being called or something..

**Comment by iViscosity** [comment]
Okay so I added a statement at the beginning of the hook to see if it's even being called correctly. It will say "OUR HOOK CALLED" in the server console whenever someone is damaged by anything. If the if statement passes, it will print "PASSED CHECKS: TARGET IS " followed by their role (such as "innocent" or "jackal"), also in the server's console (will appear blue).

**Comment by h4ppy1337** [comment]
Sorry for my late response, since I've been on vacation lately! Being back now, I can go back to focusing on our debug messages. 
Tested out the latest build and figured that the "OUR HOOK CALLED" statement indeed works as intended. Although, the Target Check does not appear. The console only prints out "OUR HOOK CALLED" (twice), followed by the regular damage info (Who damaged whom for which amount of HP)

**Comment by iViscosity** [comment]
Sorry for the delay on updates... still familiarizing myself with the new TTT2 addon function calls and working on several other projects. I have not forgotten about you :)

**Comment by h4ppy1337** [comment]
That's nice to hear, I'm counting on you :) take your time! I'll be waiting here for new updates
